### PR TITLE
Updating the Data Sources docs

### PR DESCRIPTION
**Pull request by rkoron007**
Splitting out "Fetching from REST" from a more general article covering data sources.

### CODE REVIEW

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:86]: We want to pass the Apollo Server cache to the data sources when we instantiate them. Assuming we haven't overridden the constructor in our subclass implementations, they'll share the same signature as `RESTDataSource` itself.

```ts
const { url } = await startStandaloneServer(server, {
  context: async ({ req }) => {
    const { cache } = server;
    return {
      //highlight-start
      // We create new instances of our data sources with each request
      dataSources: {
        moviesAPI: new MoviesAPI({ cache }),
        personalizationAPI: new PersonalizationAPI({ cache }),
      },
      //highlight-end
    };
  },
});
```

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:106]: Are we calling this `contextValue` everywhere or no? We might need to rename the heading, on the resolvers doc, etc.

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:131]: 👍

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:138]: We _can_ use type imports for this (though I don't know that we do that consistently in our TS docs). This is totally optional.
```suggestion
import type { KeyValueCache } from '@apollo/utils.keyvaluecache';
```

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:149]: This whole constructor can be omitted (it's the default behavior if we don't implement the constructor).

However, this also might be interesting for illustrating what's going on and what's available to your constructor. If we do leave this here, it'd be worth mentioning this happens automatically without it there.

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:172]: If this is referencing the `MoviesAPI` we created above, then we need to include the token in the constructor signature.

And if we do that, then my previous comment about dropping the default constructor is no longer valid (since `token` is not a part of the base class's constructor signature).

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [reply docs/source/data/fetching-rest.mdx:149]: See comment below in conjunction with this comment

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:182]: Custom cache for Apollo Server / external caches should definitely go on a different page from RESTDataSource stuff.

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:202]: If we're linking to a specific line number, we _shouldn't_ reference `.../blob/main/...` since that's a moving target. If the code changes on main, your link becomes invalid.

To remedy this, when you select the line in GH you'll see 3 dots - you want copy permalink. You can also just replace `main` in your URL with the most recent commit hash.

https://github.com/apollographql/datasource-rest/blob/25862e18d8b35e324c150654c5686ed317b3fca8/src/RESTDataSource.ts#L163

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:210]:
```suggestion
  override baseURL = 'https://movies-api.example.com/';
```

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:270]: The second parameter is actually an object which contains the request body (which you've correctly illustrated in your example). It also includes `method` and potentially `cacheOptions`.

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:303]: Should we just do this in our example?

File: docs/source/data/fetching-rest.mdx
Comment (trevor-scheer) [inline docs/source/data/fetching-rest.mdx:369]: We can update this to what we've been doing (remove constructor, and `override baseURL = ...`) in the class body

Comment (trevor-scheer) [reply docs/source/data/fetching-rest.mdx:369]: Pretty sure `willSendRequest` below needs an `override`

File: docs/source/data/resolvers.mdx
Comment (trevor-scheer) [inline docs/source/data/resolvers.mdx:80]: Maybe leave in or update the TODO here? Or will we have a data-sources doc?

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:383]:
```suggestion
  override baseURL = 'https://movies-api.example.com/';
```

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:385]:
```suggestion
  override willSendRequest(request: RequestOptions) {
```

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:435]:
```suggestion
  override baseURL = 'https://movies-api.example.com/';
```

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:443]:
```suggestion
  override willSendRequest(request: RequestOptions) {
```

File: docs/source/migration.mdx
Comment (trevor-scheer) [inline docs/source/migration.mdx:500]: more overrides in this class

**Review by trevor-scheer — CHANGES_REQUESTED** [review]

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:106]: Yeah, I'm going to do that in a separate PR after merging in all the AS4 docs in, so I can change `contextValue` everywhere all at once. I think it'll be less error-prone that way  😅

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:138]: Oh, rad, I'll try to do that from now on!

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:182]: Great point; I added a callout with a link to the Configuring Cache backends article. :)

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:202]: That's a great point, I always do wonder what the best way to do GH code links is. I'd rather not link folks to old commits, just in case you update things, but I can't think of a better way right now—so permalink it is!

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:270]: Oh, thank you, I forgot to correct the sentence before the example 🤦‍♀️

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:303]: Yes, yes we should!

File: docs/source/data/resolvers.mdx
Comment (rkoron007) [reply docs/source/data/resolvers.mdx:80]: I'll leave the TODO in and add a link to "Fetching from REST" for now.

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:500]: Went through and handled all these :)

File: docs/source/data/fetching-data.mdx
Comment (glasser) [inline docs/source/data/fetching-data.mdx:6]: TODO(AS4) if we want to catch before AS4 entirely done
FIXME if you don't want to merge

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:486]: I feel like this works better before the code block.

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:444]: we should probably mention somewhere that `request` now uses `@apollo/fetcher` like in `#apolloutilsfetcher-replaces-apollo-server-env` which means no `Request`/`Headers` methods

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:1621]: you could be even stronger and somehow point out that the new version has no explicit dependency on apollo-server and can be used independently of it if you want? maybe not.

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:23]: README rather than page?

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:42]: space after colon (same for next method) or just prettier it somehow

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:2]: Oh gosh. I forgot you had forked this file a little bit ago. @smyrick just added a bunch of stuff to the AS3 version of the data source docs that should probably be preserved! Added in https://github.com/apollographql/apollo-server/pull/6650 with a rename in https://github.com/apollographql/apollo-server/pull/6834

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:127]: it is probably important to mention that this cache respects headers like cache-control from the API and doesn't cache things that the backend doesn't say to cache!

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:107]: The reason this matters is that in addition to the shared cache described in the next section, there's also a SECOND per-RESTDataSource-object cache that caches *everything* regardless of HTTP headers in-memory.  Which was completely not documented until Shane's PR!  Merging his stuff in will add a bit more of that background but it would be nice if we actually described the two-level cache in the docs explicitly somewhere... see https://github.com/apollographql/apollo-server/pull/6834#pullrequestreview-1079413629

**Review by glasser — COMMENTED** [review]
looking pretty good. much of the challenge is that the existing doc page doesn't actually bother to explain the entire point of restdatasource (the fact that it's a two-level cache, one of which caches everything but only per request, and the other one respects HTTP response headers) at all...  but sorry to not make sure you knew about Shane's stuff.

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:486]: Oh, I like that; actually, I'll move this all the way up and make it the first thing in the article.

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:444]: I added a note above this code snippet:

> **New in Apollo Server 4**: Apollo Server 4 now uses [`@apollo/fetcher`](../migration#apolloutilsfetcher-replaces-apollo-server-env) under-the-hood, so you can no longer directly use `Request` or `Headers` methods.

Just to double-check, does this capture the point you are making above?

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:1621]: I like the point about removing the explicit dependency!  How about:

> - `apollo-datasource-rest` is now [`@apollo/datasource-rest`](https://www.npmjs.com/package/@apollo/datasource-rest). The `@apollo/datasource-rest` package removes an explicit dependency on `@apollo/server` and is compatible with both Apollo Server 3 and 4.

File: docs/source/migration.mdx
Comment (glasser) [reply docs/source/migration.mdx:444]: I guess, though it's more like we've actually changed the API folks write their `this.post` etc with. Compare the TS declarations:
https://www.runpkg.com/?apollo-datasource-rest@3.7.0/dist/RESTDataSource.d.ts
https://www.runpkg.com/?@apollo/datasource-rest@4.2.0/dist/RESTDataSource.d.ts

Like, the old API each of these take three args (path, body, other options) whereas in the new API it's two args (path, request where the body comes as a `body` option).

Additionally for hooks like `willSendRequest`, the object you have access to doesn't have Request/Header methods. But the bigger bit I guess is that the API changed! I think you have to adjust for `body` in the POST and PUT examples like you already did for patch.

File: docs/source/migration.mdx
Comment (glasser) [reply docs/source/migration.mdx:1621]: I guess the trick here though is that when we say "compatible with Apollo Server 3" we mean "you can use the same create-it-yourself-and-put-on-context pattern in AS3 that we document for AS4", not that you can slot it in exactly where you used the old one in AS3 (in `dataSources`).

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:127]: Great point; I added a sentence above telling folks to enable caching:

> If you've set up [caching](../preformance/caching/#recommended-starting-usage) for your server, a data source can help store the results of past fetches for particular fields in your schema.

As well as added a note after the first code snippet:

> Note that `RESTDataSource` only caches data for the fields specified by your server. To learn more about how to set up caching for your server, see [Server-side caching](../preformance/caching).

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:107]: Does it feel okay to push back documenting this to the bigger article refresh, so we can merge this faster? If not, I can get on this!

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:444]: 1. I wasn't aware the args had for the HTTP convenience methods; thank you! So I'll **remove** this paragraph because it no longer applies:
> For all methods, the **third parameter** is an `init` object that enables you to provide additional options (such as headers and referrers) to the `fetch` API that's used to send the request. For details, [see MDN's fetch docs](https://developer.mozilla.org/en-US/docs/Web/API/WindowOrWorkerGlobalScope/fetch#Parameters).

2. I'm going to add this note to the "Intercepting Fetches" section to let folks know they can't access Request/Header methods directly:

>  **New in Apollo Server 4**: Apollo Server 4 now uses [`@apollo/fetcher`](../migration#apolloutilsfetcher-replaces-apollo-server-env) under-the-hood for fetching, so hooks like `willSendRequest` no longer have direct access to `Request` or `Headers` methods.

3.  Great catch on that `POST` example, thanks!

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx]: Oh, dang, that makes so much sense....maybe I'll just remove the "compatible with Apollo Server 3" part? Unless you feel like it's valuable to keep it in? If so, I can add that background in

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [reply docs/source/data/fetching-rest.mdx:127]: That's not really it. It's that it will cache data that the *REST API* specifies as cacheable in its HTTP response headers. This has nothing to do with fields or resolvers or schema or directives, because RESTDataSource is a caching HTTP client that has literally nothing to do with GraphQL or Apollo Server except for the fact that it is one way of solving a problem that many Apollo Server users do have.

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [reply docs/source/data/fetching-rest.mdx:107]: I mean it's up to you. I feel like this sentence raises more questions than it answers though... Maybe don't even attempt to explain why you need a new instance for each operation if you're not going to provide the background for anybody to understand?

I am not sure if this is a serious suggestion or not:

> - The **`context` function should create a new instance of each `RESTDataSource` subclass for each operation.** The reason for this is related to the caching data model used by `RESTDataSource`, which is not currently documented but we intend to in the future.

File: docs/source/migration.mdx
Comment (glasser) [reply docs/source/migration.mdx]: Yeah that's probably fine.

It's compatible with Apollo Server 3 in the sense that every package on npmjs.com that is unrelated to Apollo Server is compatible with Apollo Server 3.  `apollo-datasource-rest` is a package that is very tangentially related to Apollo Server 3; `@apollo/datasource-rest` is a package that is unrelated to Apollo server 3.

File: docs/source/migration.mdx
Comment (glasser) [inline docs/source/migration.mdx:1621]: This is not technically accurate...
- `@apollo/server` is the new package name so certainly it wasn't a dependency in the past
- The thing that's removed is actually several other small package dependencies: `apollo-datasource` (the interface that Apollo Server talks to) and `apollo-server-env` (this is just us migrating to `@apollo/fetcher`.  What the point really is is that the API no longer expects to have a `configure` method called on it by Apollo Server (or by you manually?). But it's not a direct dependency.

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:127]: Okay, I obviously did not understand how that worked; thank you for explaining that! 

How does this sentence feel instead:

> The `RESTDataSource` class can cache results if the REST API it fetches from specifies caching headers in its HTTP responses (e.g., [`cache-control`](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control)).

File: docs/source/migration.mdx
Comment (rkoron007) [reply docs/source/migration.mdx:1621]: Oh sorry, I forgot to save the file before pushing up  🤦‍♀️

Hmmm, what if we remove the second sentence here? The more I think about it, with this level of detail we should either mention this in the "Fetching Rest" article directly or just omit it if we don't feel like it's important. How do you feel about that?

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:107]: Going down that road, what if we just use the first sentence for now? Omit the explanation for the moment until I can expand on this further in a (hopefully fast) follow-up:

> - The **`context` function should create a new instance of each `RESTDataSource` subclass for each operation.**

File: docs/source/migration.mdx
Comment (glasser) [reply docs/source/migration.mdx]: Yeah I think removing it is fine.

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [reply docs/source/data/fetching-rest.mdx:127]: You probably didn't understand it because we never bothered to describe what the entire point of this project is anywhere!

That sentence looks great.

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [reply docs/source/data/fetching-rest.mdx]: Yeah I think so.

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:232]:
```suggestion
      { body: { movie } }, // request body
```

Oops, this one too.

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:279]: it actually can contain a lot of stuff like `headers` too?  also missing comma after `method`

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:280]: Is that actually true? I think it's a similar object which can have `headers` and such, and params go on `params`?

File: docs/source/data/fetching-rest.mdx
Comment (glasser) [inline docs/source/data/fetching-rest.mdx:284]:
```suggestion
> **New in Apollo Server 4**: Apollo Server 4 now uses the [`@apollo/utils.fetcher`](../migration#apolloutilsfetcher-replaces-apollo-server-env) interface under the hood for fetching. This interface lets you choose your own implementation of the Fetch API. To ensure compatibility with all Fetch implementations, the request provided to hooks like `willSendRequest` is a plain JS object rather than a `Request` object with methods.
```

maybe something like this?

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:279]: Okay, you were so right—my bad for not rechecking the args! I replaced this whole section with:

> ### Method parameters
> For all HTTP convenience methods, the first parameter is the relative path of the endpoint you're sending the request to (e.g., `movies`). The second parameter is an object where you can set a request's `headers`, `params`, and `cacheOptions`, and `body` (e.g., `{ body: { id: movie.id } }`):

```ts
class MoviesAPI extends RESTDataSource {
  override baseURL = 'https://movies-api.example.com/';
  
  // an example making an HTTP POST request
  async postMovie(movie) {
    return this.post(
      `movies`, // path
      { body: movie }, // request body
    );
  }
}
```

File: docs/source/data/fetching-rest.mdx
Comment (rkoron007) [reply docs/source/data/fetching-rest.mdx:280]: (this is also addressed above)

**Review by glasser — APPROVED** [review]

File: docs/source/migration.mdx
Comment (trevor-scheer) [reply docs/source/migration.mdx:444]: For posterity, the package name is `@apollo/utils.fetcher`. (maybe you figured this out further down, I see this PR landed and I don't see any references to `@apollo/fetcher` 😄)

Comment (rkoron007) [reply docs/source/migration.mdx:444]: That did get updated, thanks for calling this out, though!

**Pull request merged**
